src/key_info_managers/sqlite_manager/mod.rs

```diff
@@ -70,28 +65,171 @@ impl SQLiteKeyInfoManager {
+                        .into_os_string()
+                        .into_string()
+                        .unwrap_or_else(|_| "DB_FILE_PATH_UNKNOWN".to_string()),
+                );
+                error!("{}", error_message);
+                return Err(Error::new(ErrorKind::Other, error_message).into());
+            }
+        }
 
-        // Create table key_mapping table if it does not already exist
-        let _ = conn.execute(
+        // The tables we require exist, check schema version matches
+        let mut schema_version_stmt = conn.prepare(
             "
-            CREATE TABLE IF NOT EXISTS key_mapping (
-                authenticator_id      INTEGER NOT NULL,
-                application_name      TEXT NOT NULL,
-                key_name              TEXT NOT NULL,
-                provider_uuid         TEXT NOT NULL,
-                provider_name         TEXT NOT NULL,
-                key_info              BLOB NOT NULL,
-                key_info_version      INTEGER NOT NULL,
-                PRIMARY KEY (authenticator_id, application_name, key_name)
-            )
-            ",
-            [],
+            SELECT
+                *
+            FROM
+                kim_metadata
+            WHERE
+                id = 'schema_version'
+        ",
         )?;
+        let mut rows = schema_version_stmt.query(params![])?;
+        while let Some(row) = rows.next()? {
+            let version_number: u8 = row.get("int_value")?;
+            if version_number != CURRENT_SCHEMA_VERSION {
+                let error_message = format!(
+                    "
+                    SQLiteKeyInfoManager database schema version is incompatible.
+                    Parsec Service is using version [{}].
+                    Database at [{}] is using version [{}].
+                    ",
+                    CURRENT_SCHEMA_VERSION,
+                    database_path
+                        .into_os_string()
+                        .into_string()
+                        .unwrap_or_else(|_| "DB_FILE_PATH_UNKNOWN".to_string()),
+                    version_number
+                );
+                error!("{}", error_message);
+                return Err(Error::new(ErrorKind::Other, error_message).into());
+            }
+        }
 
-        let mut stmt = conn.prepare("SELECT * FROM key_mapping")?;
-        let key_iter = stmt.query_map([], |row| {
+        // The tables we require exist and the schema version is the correct.
+        // Check that the key_info_version for every key is correct.
+        let mut key_info_version_stmt = conn.prepare(
+            "
+            SELECT
+                *
+            FROM
+                key_mapping
+        ",
+        )?;
+        let mut rows = key_info_version_stmt.query(params![])?;
+        while let Some(row) = rows.next()? {
+            let version_number: u8 = row.get("key_info_version")?;
+            if version_number != CURRENT_KEY_INFO_VERSION {
```

Comment:
Is it not possible (and I guess this applies to the previous query as well) to just expand the query to ask for entries where the version number has this property (i.e. `!= CURRENT_KEY_INFO_VERSION`)? Would've thought SQL allows us to do that, and then you can just check the row count.

Reply:
Yes 🤦🏼‍♂️ think I was just on autopilot copying from one of the previous SQL statements.

Reply:
Having looked at it a second time. I've added the `WHERE key_info_version != CURRENT_KEY_INFO_VERSION` clause to the SQL statement. I've however chosen not to use the length, this is as the value returned from the query only implements iterator. I would have to call `.count()` which under the hood calls `.next()` for every item.

Opted for this instead:
```
let rows = SQL_QUERY();
// If a mapping exists with the wrong key_info_version, throw an error.
if let Some(row) = rows.next()? {
    .....throw error
}
```